File: shared/src/graph/InstantiatedGraph.test.ts
import { initInstantiatedGraph, InstantiatedGraph, StreamIdOverridesForRootSpec } from "./InstantiatedGraph";
import * as DefGraphModule from "./DefGraph";
import { initDefGraph } from "./DefGraph";

/**
 * Removed the FakeDefGraph class.
 * Originally a fake defGraph was defined here to support the test; it has been removed in favor of a real defGraph.
 */

// For testing purposes we override the asynchronous genSpecIdentifier.
// In our InstantiatedGraph code the spec job id is computed as:
//    `[${contextId}]${await genSpecIdentifier(specName, uniqueSpecLabel)}`
jest.spyOn(DefGraphModule, "genSpecIdentifier").mockImplementation(async (name: string, uniqueLabel?: string) => {
  // Simply return the spec name so that job id becomes `[testContext]SpecA`
  return name;
});

describe("InstantiatedGraph", () => {
  it("should instantiate from a defGraph and create all nodes and edges properly", async () => {
    // Create a real DefGraph.
    const realDefGraph = initDefGraph({
      root: {
        name: "RootSpec",
        inputTags: [],
        outputTags: [],
      },
    });
    
    // Add a SpecA node using ensureInletAndStream.
    // This call creates (if needed) a Spec node with specName 'SpecA'
    // along with an inlet node (using a dummy tag 'dummy') and a new stream node.
    (realDefGraph as any).ensureInletAndStream({ specName: "SpecA", tag: "dummy", uniqueSpecLabel: null }, false);

    // Create the InstantiatedGraph passing in our real defGraph and other needed properties.
    const instantiatedGraph = initInstantiatedGraph({
      contextId: "testContext",
      defGraph: realDefGraph as any, // cast to any (or DefGraph) as needed
      streamIdOverrides: {} as StreamIdOverridesForRootSpec,
      rootJobId: "rootJob",
      inletHasTransformOverridesByTag: {},
      streamSourceSpecTypeByStreamId: {},
    });


    // Retrieve all node IDs from the instantiated graph.
    const nodeIds = instantiatedGraph.nodes();

    // Verify that one node has the jobId "rootJob" (mapped from the RootSpec).
    const rootJobNodeFound = nodeIds.some((id) => {
      const node = instantiatedGraph.getNodeAttributes(id);
      return 'jobId' in node && node.jobId === "rootJob";
    });

    // Verify that one node has the jobId "[testContext]SpecA" (mapped from the Spec node).
    const specJobNodeFound = nodeIds.some((id) => {
      const node = instantiatedGraph.getNodeAttributes(id);
      return 'jobId' in node && node.jobId === "[testContext]SpecA";
    });

    expect(rootJobNodeFound).toBeTruthy();
    expect(specJobNodeFound).toBeTruthy();

    // Check that the number of edges in the instantiated graph equals the number of edges in realDefGraph.
    const instantiatedEdgeCount = instantiatedGraph.edgeCount();
    const defGraphEdgeCount = (realDefGraph).edges().results.length;
    expect(instantiatedEdgeCount).toBe(defGraphEdgeCount);
  });

  it("should instantiate with multiple spec nodes and compute correct job ids", async () => {
    // Create a real DefGraph.
    const realDefGraph = initDefGraph({
      root: {
        name: "RootSpec",
        inputTags: [],
        outputTags: [],
      },
    });
    // Add two Spec nodes: SpecA and SpecB using ensureInletAndStream.
    (realDefGraph).ensureInletAndStream({ specName: "SpecA", tag: "dummyA", uniqueSpecLabel: null }, false);
    (realDefGraph).ensureInletAndStream({ specName: "SpecB", tag: "dummyB", uniqueSpecLabel: null }, false);

    const instantiatedGraph = initInstantiatedGraph({
      contextId: "multiTest",
      defGraph: realDefGraph,
      streamIdOverrides: {} as StreamIdOverridesForRootSpec,
      rootJobId: "rootJob",
      inletHasTransformOverridesByTag: {},
      streamSourceSpecTypeByStreamId: {},
    });

    const nodeIds = instantiatedGraph.nodes();
    const foundSpecA = nodeIds.some((id) => {
      const node = instantiatedGraph.getNodeAttributes(id);
      return 'jobId' in node && node.jobId === "[multiTest]SpecA";
    });
    const foundSpecB = nodeIds.some((id) => {
      const node = instantiatedGraph.getNodeAttributes(id);
      return 'jobId' in node && node.jobId === "[multiTest]SpecB";
    });

    expect(foundSpecA).toBeTruthy();
    expect(foundSpecB).toBeTruthy();
  });

  it("should have the same node count as the underlying defGraph", async () => {
    const realDefGraph = initDefGraph({
      root: {
        name: "RootSpec",
        inputTags: ["in1"],
        outputTags: ["out1"],
      },
    });
    (realDefGraph).ensureInletAndStream({ specName: "SpecA", tag: "dummy", uniqueSpecLabel: null }, false);

    const instantiatedGraph = initInstantiatedGraph({
      contextId: "nodeCountTest",
      defGraph: realDefGraph,
      streamIdOverrides: {} as StreamIdOverridesForRootSpec,
      rootJobId: "rootJob",
      inletHasTransformOverridesByTag: {},
      streamSourceSpecTypeByStreamId: {},
    });

    const instantiatedNodeCount = instantiatedGraph.nodes().length;
    const defGraphNodeCount = realDefGraph.nodes().length;
    expect(instantiatedNodeCount).toBe(defGraphNodeCount);
  });

  it("should retrieve correct node attributes for all nodes", async () => {
    const realDefGraph = initDefGraph({
      root: {
        name: "RootSpec",
        inputTags: ["in1"],
        outputTags: ["out1"],
      },
    });
    (realDefGraph).ensureInletAndStream({ specName: "SpecA", tag: "dummy", uniqueSpecLabel: null }, false);

    const instantiatedGraph = initInstantiatedGraph({
      contextId: "attrTest",
      defGraph: realDefGraph,
      streamIdOverrides: {} as StreamIdOverridesForRootSpec,
      rootJobId: "rootJob",
      inletHasTransformOverridesByTag: {},
      streamSourceSpecTypeByStreamId: {},
    });

    const nodeIds = instantiatedGraph.nodes();
    nodeIds.forEach((id) => {
      const attr = instantiatedGraph.getNodeAttributes(id);
      expect(attr).toHaveProperty("label");
      // For Spec and RootSpec nodes, check that the jobId property is defined.
      if (attr.nodeType === "job" || attr.nodeType === "root-job") {
        expect(attr).toHaveProperty("jobId");
      }
    });
  });
}); 

File: shared/tests/instantiated_graph_test.rs
use std::collections::HashMap;
use livestack_shared::systems::def_graph::{DefGraph, DefGraphNode, DefGraphNodeType};
use livestack_shared::systems::instantiated_graph::InstantiatedGraph;

#[cfg(test)]
mod tests {
    use super::*;

    /// This test corresponds to "should instantiate from a defGraph and create all nodes and edges properly"
    /// in the original TypeScript test. It verifies that nodes and edges are properly created when instantiated.
    #[test]
    fn should_instantiate_from_a_def_graph_and_create_all_nodes_and_edges_properly() {
        let root_spec_name = "RootSpec".to_string();
        let input_tags = vec!["input1".to_string(), "input2".to_string()];
        let output_tags = vec!["output1".to_string(), "output2".to_string()];
        let mut def_graph = DefGraph::new(root_spec_name.clone(), input_tags, output_tags);

        // Add a spec node to the DefGraph
        let spec_node_data = DefGraphNode {
            node_type: DefGraphNodeType::Spec,
            spec_name: Some("SpecA".to_string()),
            unique_spec_label: None,
            tag: None,
            has_transform: None,
            stream_def_id: None,
            alias: None,
            direction: None,
            label: "SpecA".to_string(),
        };
        let _spec_node_id = def_graph.ensure_node("SpecA", spec_node_data);

        // Instantiate an InstantiatedGraph from the DefGraph
        let instantiated_graph = InstantiatedGraph::new(
            "testContext".to_string(),
            "rootJob".to_string(),
            HashMap::new(), // stream_id_overrides
            HashMap::new(), // inlet_has_transform_overrides_by_tag
            HashMap::new(), // stream_source_spec_type_by_stream_id
            &def_graph,
        );

        // Check that a node has job_id "rootJob" (for the RootSpec).
        // Also check that another has job_id "[testContext]SpecA".
        let node_indices = instantiated_graph.node_indices();
        let root_job_found = node_indices.iter().any(|node_id| {
            instantiated_graph.node_weight(*node_id)
                .and_then(|node| node.job_id.as_ref())
                == Some(&"rootJob".to_string())
        });
        let spec_a_found = node_indices.iter().any(|node_id| {
            instantiated_graph.node_weight(*node_id)
                .and_then(|node| node.job_id.as_ref())
                == Some(&"[testContext]SpecA".to_string())
        });
        assert!(root_job_found, "Expected a node with job_id == 'rootJob'.");
        assert!(spec_a_found, "Expected a node with job_id == '[testContext]SpecA'.");

        // Check edge counts match
        let edge_count_in_instantiated_graph = instantiated_graph.edge_count();
        let edge_count_in_def_graph = def_graph.edge_count();
        assert_eq!(
            edge_count_in_instantiated_graph, edge_count_in_def_graph,
            "InstantiatedGraph and DefGraph should have the same number of edges"
        );
    }

    /// This test corresponds to "should instantiate with multiple spec nodes and compute correct job ids"
    /// in the original TypeScript test. It checks whether multiple specs get correct job IDs in the instantiated graph.
    #[test]
    fn should_instantiate_with_multiple_spec_nodes_and_compute_correct_job_ids() {
        let mut def_graph = DefGraph::new(
            "RootSpec".to_string(),
            vec![],  // no input tags
            vec![],  // no output tags
        );

        // Add two spec nodes (SpecA, SpecB) with inlets so they appear as specs in def_graph
        def_graph.ensure_inlet_and_stream(
            livestack_shared::systems::def_graph_utils::SpecTagInfo {
                spec_name: "SpecA".to_string(),
                tag: "dummyA".to_string(),
                unique_spec_label: None,
            },
            false,
        );
        def_graph.ensure_inlet_and_stream(
            livestack_shared::systems::def_graph_utils::SpecTagInfo {
                spec_name: "SpecB".to_string(),
                tag: "dummyB".to_string(),
                unique_spec_label: None,
            },
            false,
        );

        // Instantiate
        let instantiated_graph = InstantiatedGraph::new(
            "multiTest".to_string(),
            "rootJob".to_string(),
            HashMap::new(), // stream_id_overrides
            HashMap::new(), // inlet_has_transform_overrides_by_tag
            HashMap::new(), // stream_source_spec_type_by_stream_id
            &def_graph,
        );

        let node_indices = instantiated_graph.node_indices();
        let found_spec_a = node_indices.iter().any(|id| {
            let node = instantiated_graph.node_weight(*id).unwrap();
            node.job_id.as_deref() == Some("[multiTest]SpecA")
        });
        let found_spec_b = node_indices.iter().any(|id| {
            let node = instantiated_graph.node_weight(*id).unwrap();
            node.job_id.as_deref() == Some("[multiTest]SpecB")
        });

        assert!(found_spec_a, "Expected a node with job_id == '[multiTest]SpecA'.");
        assert!(found_spec_b, "Expected a node with job_id == '[multiTest]SpecB'.");
    }

    /// This test corresponds to "should have the same node count as the underlying defGraph"
    /// in the original TypeScript test. It verifies that InstantiatedGraph has the same number of nodes as DefGraph.
    #[test]
    fn should_have_the_same_node_count_as_the_underlying_def_graph() {
        let mut def_graph = DefGraph::new(
            "RootSpec".to_string(),
            vec!["in1".to_string()],
            vec!["out1".to_string()],
        );

        def_graph.ensure_inlet_and_stream(
            livestack_shared::systems::def_graph_utils::SpecTagInfo {
                spec_name: "SpecA".to_string(),
                tag: "dummy".to_string(),
                unique_spec_label: None,
            },
            false,
        );

        let instantiated_graph = InstantiatedGraph::new(
            "nodeCountTest".to_string(),
            "rootJob".to_string(),
            HashMap::new(),
            HashMap::new(),
            HashMap::new(),
            &def_graph,
        );

        let instantiated_node_count = instantiated_graph.node_indices().len();
        let def_graph_node_count = def_graph.node_count();

        assert_eq!(
            instantiated_node_count, def_graph_node_count,
            "InstantiatedGraph should have the same node count as the underlying DefGraph"
        );
    }

    /// This test corresponds to "should retrieve correct node attributes for all nodes"
    /// in the original TypeScript test. Here, we check that every node has a label, and that
    /// job/root-job nodes also have a jobId.
    #[test]
    fn should_retrieve_correct_node_attributes_for_all_nodes() {
        let mut def_graph = DefGraph::new(
            "RootSpec".to_string(),
            vec!["in1".to_string()],
            vec!["out1".to_string()],
        );
        def_graph.ensure_inlet_and_stream(
            livestack_shared::systems::def_graph_utils::SpecTagInfo {
                spec_name: "SpecA".to_string(),
                tag: "dummy".to_string(),
                unique_spec_label: None,
            },
            false,
        );

        let instantiated_graph = InstantiatedGraph::new(
            "attrTest".to_string(),
            "rootJob".to_string(),
            HashMap::new(),
            HashMap::new(),
            HashMap::new(),
            &def_graph,
        );

        for node_id in instantiated_graph.node_indices() {
            let node = instantiated_graph.node_weight(node_id).unwrap();
            // The node must have some label
            assert!(
                !node.label.is_empty(),
                "Expected the node to have a non-empty label"
            );

            // If it's a job or root-job type, then job_id should exist
            match node.node_type {
                livestack_shared::systems::instantiated_graph::InstantiatedNodeType::RootJob
                | livestack_shared::systems::instantiated_graph::InstantiatedNodeType::Job => {
                    assert!(
                        node.job_id.is_some(),
                        "Expected job_id to be present for a job or root-job node"
                    );
                }
                _ => {
                    // Other node types won't have a job_id, so we don't check them here
                }
            }
        }
    }
} 